validator/stateless_block_validator.go

```diff
@@ -364,90 +405,105 @@ func RecordBlockCreation(
 
 	var preimages map[common.Hash][]byte
 	if recordingKV != nil {
-		preimages, err = arbitrum.PreimagesFromRecording(chaincontext, recordingKV)
+		preimages, err = v.recordingDatabase.PreimagesFromRecording(chaincontext, recordingKV)
 		if err != nil {
 			return common.Hash{}, nil, nil, err
 		}
 	}
+	if keepReference && err == nil {
+		prevHeader = nil
+	}
 	return blockHash, preimages, readBatchInfo, err
 }
 
-func BlockDataForValidation(
-	ctx context.Context,
-	blockchain *core.BlockChain,
-	inboxReader InboxReaderInterface,
-	header, prevHeader *types.Header,
-	msg arbstate.MessageWithMetadata,
-	producePreimages bool,
-) (
-	preimages map[common.Hash][]byte, readBatchInfo []BatchInfo,
-	hasDelayedMessage bool, delayedMsgNr uint64, err error,
-) {
-	var prevHash common.Hash
-	if prevHeader != nil {
-		prevHash = prevHeader.Hash()
+func (v *StatelessBlockValidator) ValidationEntryRecord(ctx context.Context, e *validationEntry, keepReference bool) error {
+	if e.Stage != ReadyForRecord {
+		return errors.Errorf("validation entry should be ReadyForRecord, is: %v", e.Stage)
 	}
-	if header.ParentHash != prevHash {
-		err = fmt.Errorf("bad arguments: prev does not match")
-		return
+	if e.PrevBlockHeader == nil {
+		e.Stage = Recorded
+		return nil
 	}
-
-	if prevHeader != nil {
-		var blockhash common.Hash
-		blockhash, preimages, readBatchInfo, err = RecordBlockCreation(
-			ctx, blockchain, inboxReader, prevHeader, &msg, producePreimages,
-		)
+	blockhash, preimages, readBatchInfo, err := v.RecordBlockCreation(ctx, e.PrevBlockHeader, e.msg, keepReference)
+	if err != nil {
+		return err
+	}
+	if blockhash != e.BlockHash {
+		return fmt.Errorf("recording failed: blockNum %d, hash expected %v, got %v", e.BlockNumber, e.BlockHash, blockhash)
+	}
+	if e.HasDelayedMsg {
+		delayedMsg, err := v.inboxTracker.GetDelayedMessageBytes(e.DelayedMsgNr)
 		if err != nil {
-			return
-		}
-		if blockhash != header.Hash() {
-			err = fmt.Errorf("wrong hash expected %s got %s", header.Hash(), blockhash)
-			return
+			log.Error(
+				"error while trying to read delayed msg for proving",
+				"err", err, "seq", e.DelayedMsgNr, "blockNr", e.BlockNumber,
+			)
+			return errors.New("error while trying to read delayed msg for proving")
```

Comment:
```suggestion
			return fmt.Errorf("error while trying to read delayed msg for proving: %w", err)
```
We should use fmt.Errorf to wrap the inner error here